apps/wallet/src/background/account-sources/MnemonicAccountSource.ts

```diff
@@ -0,0 +1,172 @@
+			sourceHash: bytesToHex(sha256(entropy)),
+		};
+		const allAccountSources = await getAccountSources();
+		for (const anAccountSource of allAccountSources) {
+			if (
+				anAccountSource instanceof MnemonicAccountSource &&
+				(await anAccountSource.sourceHash) === dataSerialized.sourceHash
+			) {
+				throw new Error('Mnemonic account source already exists');
+			}
+		}
+		await (await getDB()).accountSources.put(dataSerialized);
+		await backupDB();
+		accountSourcesEvents.emit('accountSourcesChanged');
+		return new MnemonicAccountSource(dataSerialized.id);
+	}
+
+	static isOfType(
+		serialized: AccountSourceSerialized,
+	): serialized is MnemonicAccountSourceSerialized {
+		return serialized.type === 'mnemonic';
+	}
+
+	constructor(id: string) {
+		super({ type: 'mnemonic', id });
+	}
+
+	async isLocked() {
+		return (await this.getEphemeralValue()) === null;
+	}
+
+	async unlock(password: string) {
+		await this.setEphemeralValue(
+			await decrypt(password, (await this.getStoredData()).encryptedData),
+		);
+		accountSourcesEvents.emit('accountSourceStatusUpdated', { accountSourceID: this.id });
+	}
+
+	async lock() {
+		await this.clearEphemeralValue();
+		accountSourcesEvents.emit('accountSourceStatusUpdated', { accountSourceID: this.id });
+	}
+
+	async deriveAccount(): Promise<Omit<MnemonicSerializedAccount, 'id'>> {
+		const derivationPath = await this.getAvailableDerivationPath();
+		const keyPair = await this.deriveKeyPair(derivationPath);
+		return {
+			type: 'mnemonic-derived',
+			sourceID: this.id,
+			address: keyPair.getPublicKey().toSuiAddress(),
+			derivationPath,
+			publicKey: keyPair.getPublicKey().toBase64(),
+			lastUnlockedOn: null,
+		};
+	}
+
+	async deriveKeyPair(derivationPath: string) {
+		const data = await this.getEphemeralValue();
+		if (!data) {
+			throw new Error('Locked');
+		}
+		return Ed25519Keypair.deriveKeypairFromSeed(data.mnemonicSeedHex, derivationPath);
+	}
+
+	async toUISerialized(): Promise<MnemonicAccountSourceSerializedUI> {
+		const { type } = await this.getStoredData();
+		return {
+			id: this.id,
+			type,
+			isLocked: await this.isLocked(),
+		};
+	}
+
+	get sourceHash() {
+		return this.getStoredData().then(({ sourceHash }) => sourceHash);
+	}
+
+	private makeDerivationPath(index: number) {
+		// currently returns only Ed25519 path
+		return `m/44'/784'/${index}'/0'/0'`;
+	}
+
+	private async getAvailableDerivationPath() {
+		const derivationPathMap: Record<string, boolean> = {};
+		for (const anAccount of await getAllAccounts({ sourceID: this.id })) {
+			if (anAccount instanceof MnemonicAccount && (await anAccount.sourceID) === this.id) {
+				derivationPathMap[await anAccount.derivationPath] = true;
+			}
+		}
+		let index = 0;
+		let derivationPath = '';
+		let temp;
+		do {
+			temp = this.makeDerivationPath(index++);
+			if (!derivationPathMap[temp]) {
+				derivationPath = temp;
+			}
+		} while (derivationPath === '' && index < 10000);
```

Comment:
I added this for some kind of safety in case something goes wrong we won't get in an infinite loop. And chose 10000 as big enough that wont limit users from creating accounts and not very big to be noticeable for any error/infinite loop case